kore/src/Kore/Step/Axiom/Matcher.hs

```diff
@@ -9,93 +9,494 @@ Stability   : experimental
+import qualified Kore.Internal.TermLike as TermLike
+import           Kore.Predicate.Predicate
+                 ( makeCeilPredicate )
+import qualified Kore.Predicate.Predicate as Syntax
+                 ( Predicate )
+import qualified Kore.Predicate.Predicate as Syntax.Predicate
 import           Kore.Step.Simplification.AndTerms
                  ( SortInjectionMatch (SortInjectionMatch),
                  simplifySortInjections )
 import qualified Kore.Step.Simplification.AndTerms as SortInjectionMatch
                  ( SortInjectionMatch (..) )
 import qualified Kore.Step.Simplification.AndTerms as SortInjectionSimplification
                  ( SortInjectionSimplification (..) )
-import qualified Kore.Step.Simplification.Ceil as Ceil
-                 ( makeEvaluateTerm )
 import qualified Kore.Step.Simplification.Data as Simplifier
-import           Kore.Step.Substitution
-                 ( createPredicatesAndSubstitutionsMergerExcept,
-                 mergePredicatesAndSubstitutionsExcept )
 import           Kore.Unification.Error
-                 ( unsupportedPatterns )
+                 ( SubstitutionError (..), unsupportedPatterns )
 import           Kore.Unification.Procedure
                  ( unificationProcedure )
+import qualified Kore.Unification.Substitution as Substitution
 import           Kore.Unification.Unify
                  ( MonadUnify )
 import qualified Kore.Unification.Unify as Monad.Unify
 import           Kore.Unparser
 import           Kore.Variables.Fresh
                  ( FreshVariable )
 
-{- Matches two patterns based on their form.
+-- * Matching
 
-Assumes that the two patterns have no common variables (quantified or not).
+{- | Dispatch a single matching constraint.
 
-Returns Right bottom or Left when it can't handle the patterns. The
-returned substitution substitutes only variables from the first pattern.
+@matchOne@ is the heart of the matching algorithm. Each matcher is applied to
+the constraint in sequence, until one accepts the pair. The matchers may
+introduce substitutions and new constraints. If none of the matchers accepts the
+pair, it is deferred until we have more information.
 
-The meaning of a Right value is that the substitution holds IF the predicate
-holds.
+See also: 'push', 'substitute', 'defer'.
 
-TODO: This is different from unification's meaning, so we should either
-convert all bottoms to Left, or we should do it selectively. Doing
-it selectively is not trivial, e.g. a bottom inside a function should become
-Left, but inside a constructor we may be able to keep it as bottom.
--}
-matchAsUnification
-    ::  ( FreshVariable variable
-        , Show variable
-        , Unparse variable
-        , SortedVariable variable
-        , MonadUnify unifier
-        )
+ -}
+matchOne
+    :: (MatchingVariable variable, MonadUnify unifier)
+    => Pair (TermLike variable)
+    -> MatcherT variable unifier ()
+matchOne pair =
+    (   matchEqual       pair
```

Comment:
I was planning to implement matching for (at least) `\exists` and `\equals` because my next task depends on it, so I didn't want to cover all the pattern heads, but this covers domain values, literals, and builtins. And if two patterns are literally equal, then they must match. But, I should probably single out the cases that we have tests for right now.